### PR TITLE
Support path with uuid

### PR DESCRIPTION
**Pull request by masterix21**
It solves the problems of UUID change using the `TemporaryUpload` model from laravel-medialibrary-pro.

### CODE REVIEW

**Comment by freekmurze** [comment]
I very much appreciate your help, but I'm not going to merge this in.

I think this adds complexity for something the vast majority of users don't need. 

You can add these changes to your own project, by using a custom Media model and applying these changes there.

**Comment by masterix21** [comment]
I trust you, but I think the UUID part isn't only a part of my own project: it's a part of the package itself. Also, I think UUID should be unique and only one time generated for a model, and the part of package code that regenerates the UUID is no-sense.

If you agree with me, I will remove the UUID regeneration from your package, but before, I'd like to hear your opinion to avoid working on a refused PR. 

Thanks.

**Comment by freekmurze** [comment]
Reconsidering this one...

File: src/MediaCollections/Models/Concerns/PathUuidSafe.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
+<?php
+
+namespace Spatie\MediaLibrary\MediaCollections\Models\Concerns;
+
+use Illuminate\Support\Facades\Storage;
+use Spatie\MediaLibrary\MediaCollections\Models\Media;
+use Spatie\MediaLibrary\Support\PathGenerator\PathGeneratorFactory;
+
+trait PathUuidSafe
```
Comment (freekmurze) [inline src/MediaCollections/Models/Concerns/PathUuidSafe.php:9]: Please rename this to `SupportsUuidInPaths`

File: src/MediaCollections/Models/Concerns/PathUuidSafe.php [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+{
+    public static function bootPathUuidSafe()
+    {
+        if (! config('media-library.path_uses_uuid', false)) {
```
Comment (freekmurze) [inline src/MediaCollections/Models/Concerns/PathUuidSafe.php:13]: The `false` default can be removed, the config value is always present.

File: src/MediaCollections/Models/Concerns/PathUuidSafe.php [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+            return;
+        }
+
+        static::updating(function (Media $media) {
```
Comment (freekmurze) [inline src/MediaCollections/Models/Concerns/PathUuidSafe.php:17]: Could you add a test for this behaviour?

**Review by freekmurze — CHANGES_REQUESTED** [review]
Let's make a few final polishments before merging it in.

**Comment by masterix21** [comment]
What do you think if I rename `media-library.path_uses_uuid` to `media-library.rename_path_on_media_saving`, and obviously, I rename all other parts of code?

Because I think it could be used any time, for any custom PathGenerator, for who uses the UUID or for who uses the collection, or any other part of the model that could change in time.

What's your idea?

**Comment by freekmurze** [comment]
How about `moves_media_on_update`?

**Comment by masterix21** [comment]
It sounds good.

**Comment by masterix21** [comment]
or these?
- config `keep_consistent_path`
- trait `HasConsistentPath`

**Comment by freekmurze** [comment]
Good suggestions, but I think I like `moves_media_on_update` most.

**Comment by masterix21** [comment]
Ok 😉

File: src/MediaCollections/Models/Concerns/PathUuidSafe.php (hunk 1 above)
Comment (masterix21) [reply src/MediaCollections/Models/Concerns/PathUuidSafe.php]: Renamed to `MovesOnUpdate`. Is it ok?

File: src/MediaCollections/Models/Concerns/PathUuidSafe.php (hunk 2 above)
Comment (masterix21) [reply src/MediaCollections/Models/Concerns/PathUuidSafe.php]: Ok

File: src/MediaCollections/Models/Concerns/PathUuidSafe.php (hunk 3 above)
Comment (masterix21) [reply src/MediaCollections/Models/Concerns/PathUuidSafe.php]: Yes, sure.

**Comment by masterix21** [comment]
@freekmurze, please don't hate me. I have removed the trait and moved the code inside the observer, so I created a method inside the filesystem file.

**Comment by freekmurze** [comment]
> please don't hate me. 

Never! ❤️

File: src/MediaCollections/Filesystem.php [hunk 4]
```diff
@@ -229,6 +229,16 @@ public function syncFileNames(Media $media): void
         $this->renameConversionFiles($media);
     }
 
+    public function syncMediaPath(Media $media): void
+    {
+        $factory = PathGeneratorFactory::create();
+
+        $oldMedia = (clone $media)->fill($media->getOriginal());
+
+        $this->filesystem->disk($media->disk)
```
Comment (freekmurze) [inline src/MediaCollections/Filesystem.php:238]: Maybe we should compare the old path and with the new path before calling `move`

**Review by freekmurze — CHANGES_REQUESTED** [review]
Looks good. I do see one improvement.

File: src/MediaCollections/Filesystem.php (hunk 4 above)
Comment (masterix21) [reply src/MediaCollections/Filesystem.php:238]: Yes, it's good. So, we can remove the config `moves_media_on_update` and call the sync method when the paths are not equals. Is it right?

Comment (freekmurze) [reply src/MediaCollections/Filesystem.php:242]: I would still only do the move if the `moves_media_on_update` is set to `true`. 

The base idea is that after upload, medialibrary won't any files, unless you explicitly enable it. Feel safer to me!

**Review by freekmurze — APPROVED** [review]

**Comment by freekmurze** [comment]
Thanks!

**Comment by lostincode** [comment]
@masterix21 @freekmurze So this solution is great but will not work with `s3` disk since the filesystem cannot move an entire directory. (took me awhile to figure that one out).

The universal solution for all disks is something like this:

```php

    public function updating(Media $media)
    {
        $oldMedia = (clone $media)->fill($media->getOriginal());

        if ($oldMedia->getPath() === $media->getPath()) {
            return;
        }

        $factory = PathGeneratorFactory::create();

        $oldPath = $factory->getPath($oldMedia);
        $newPath = $factory->getPath($media);

        $images = Storage::disk('s3')->allFiles($oldPath);

        foreach ($images as $img) {
            $moveTo = str_replace($oldPath, $newPath, $img);
            Storage::disk('s3')->move($img, $moveTo);
        }
    }

```

Since with `s3` disk you can only move a single file at a time and not the full directory. It seems like not many people use the `s3` driver with UUID in path so I'm just sharing a fix for anyone using this combo like me 😬

**Comment by masterix21** [comment]
@lostincode great, thanks for your share.

**Comment by dhcmega** [comment]
Hi
I just spent a few hours trying to understand why this didn't work:

```
    protected function getBasePath(Media $media): string
    {
        return "media-library/{$account_id}/{$media->getKey()}/{$media->uuid}";
    }
```

It was because the model gets created with a different UUID than the one it received. But the file gets saved using the original one. When it tries to move it, the uuid is different (don't know why) and fails to find the file.
Also noticed that `getPath()` returns the path with `/` at the end and it is used in `Filesystem::getStream()` with an extra `/` when concatenating.

I don't like the idea of using `md5($media->id)` because every time the file is read, it will have to recalculate de md5.

Thanks

**Comment by greeto** [comment]
What was the solution here?
users see this /media/2/my-image.png

I would like them to see /media/UUID/my-image.png

**Comment by hakobfromlightin** [comment]
Seems like there is no solution for this bug. When using storage, the files are not moved. This is clearly a bug.

**Pull request merged**